### PR TITLE
Apache Arrow native reads for JS+Python

### PR DESCRIPTION
**Pull request by texodus**
Ports Apache Arrow reading to use the C++ (rather than the TS) version of the library.  In addition to significant load time performance improvements, this makes arrow loading consistent & portable between Python & JS.

### CODE REVIEW

**Comment by timkpaine** [comment]
Don't need the rapidjson cmake file

File: cmake/arrow/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+target_link_libraries(arrow
+    ${double-conversion_LIBRARIES}
+    ${BOOST_FILESYSTEM_LIBRARY}
+    ${BOOST_SYSTEM_LIBRARY}
+    ${ARROW_TEST_LINK_TOOLCHAIN})
+
+add_custom_target(arrow_fb_files
+                  # COMMAND node $<TARGET_FILE:flatc>
+                  COMMAND /usr/local/bin/flatc
```
Comment (timkpaine) [inline cmake/arrow/CMakeLists.txt]: Don't hardcore path

File: cpp/perspective/CMakeLists.txt [hunk 2]
```diff
@@ -154,9 +170,10 @@ if (PSP_WASM_BUILD)
 	set(CMAKE_RANLIB emranlib)
 	set(CMAKE_EXECUTABLE_SUFFIX ".js")
 	list(APPEND CMAKE_PREFIX_PATH /usr/local)
-	include_directories("/usr/local/include")
 	include_directories("/usr/local/Homebrew/include")
```
Comment (timkpaine) [inline cpp/perspective/CMakeLists.txt]: Mac only

File: cpp/perspective/CMakeLists.txt [hunk 3]
```diff
@@ -154,9 +170,10 @@ if (PSP_WASM_BUILD)
...
+	include_directories("/usr/local/include" SYSTEM)
 	include_directories("/boost_includes")
+	include_directories("/usr/local/flatbuffers/include")
```
Comment (timkpaine) [inline cpp/perspective/CMakeLists.txt]: Don't hardcode

File: cpp/perspective/CMakeLists.txt [hunk 4]
```diff
@@ -212,7 +232,9 @@ elseif(PSP_CPP_BUILD OR PSP_PYTHON_BUILD)
 	set(ASYNC_MODE_FLAGS "")
 	set(ASMJS_MODE_FLAGS "")
 
-	find_package(Boost)
+	include_directories("/usr/local/flatbuffers/include")
```
Comment (timkpaine) [inline cpp/perspective/CMakeLists.txt]: Don't hardcode

File: cpp/perspective/CMakeLists.txt [hunk 5]
```diff
@@ -257,17 +279,46 @@ elseif(PSP_CPP_BUILD OR PSP_PYTHON_BUILD)
...
+			message(FATAL_ERROR "${Red}PyArrow could not be located${ColorReset}")
+		else()
+			message(WARNING "${Cyan}PyArrow found: ${PYTHON_PYARROW_INCLUDE_DIR}${ColorReset}")
+		endif()
+		include_directories(${PYTHON_PYARROW_INCLUDE_DIR})
+		link_directories(${PYTHON_PYARROW_LIBRARY_DIR})
+
+		message("${Cyan}Building minimal Apache Arrow${ColorReset}")
+		# psp_build_dep("double-conversion" "../../cmake/double-conversion.txt.in")
+		# psp_build_dep("rapidjson" "../../cmake/rapidjson.txt.in")
```
Comment (timkpaine) [inline cpp/perspective/CMakeLists.txt]: Remove

File: cpp/perspective/CMakeLists.txt [hunk 6]
```diff
@@ -257,17 +279,46 @@ elseif(PSP_CPP_BUILD OR PSP_PYTHON_BUILD)
...
+
 		#####################
 	endif()
 endif()
 
 psp_build_dep("hopscotch" "../../cmake/hopscotch.txt.in")
 psp_build_dep("ordered-map" "../../cmake/ordered-map.txt.in")
+psp_build_dep("double-conversion" "../../cmake/double-conversion.txt.in")
+# psp_build_dep("flatbuffers" "../../cmake/flatbuffers.txt.in")
```
Comment (timkpaine) [inline cpp/perspective/CMakeLists.txt]: Remove

File: cpp/perspective/CMakeLists.txt [hunk 7]
```diff
@@ -370,13 +422,16 @@ set (PYTHON_SOURCE_FILES
 if (WIN32)
 	set(CMAKE_CXX_FLAGS " /EHsc")
 else()
-	set(CMAKE_CXX_FLAGS " -std=c++0x ${CMAKE_CXX_FLAGS} ${CMAKE_C_FLAGS}")
+	set(CMAKE_CXX_FLAGS " ${CMAKE_CXX_FLAGS} ${CMAKE_C_FLAGS}")
 endif()
 
 if (PSP_WASM_BUILD)
```
Comment (timkpaine) [inline cpp/perspective/CMakeLists.txt]: Did we decide to build rapidjson?

File: docs/md/development.md [hunk 8]
```diff
@@ -68,11 +68,11 @@ Installing and activating the latest [emscripten SDK](https://github.com/kripken
 ./emsdk activate latest
 ```
 
-You'll also need Boost and CMake, which can be installed from Homebrew:
+You'll also need FlatBuffers and CMake, which can be installed from Homebrew:
 
 ```bash
 brew install cmake
```
Comment (timkpaine) [inline docs/md/development.md]: Also rapidjson

File: cpp/perspective/CMakeLists.txt (hunk 7 above)
Comment (texodus) [reply cpp/perspective/CMakeLists.txt]: Yes.

**Comment by wesm** [comment]
You're probably going to be interested in https://issues.apache.org/jira/browse/ARROW-6637

Flatbuffers is no longer required to build the project anymore. We're working on eliminating Boost and other currently-mandatory build dependencies.

**Comment by timkpaine** [comment]
@wesm thanks, flatbuffers was a pain to figure out but it's not so bad now since we use the system flatbuffers to both generate the files and compile headers under wasm. We're using a pretty small cut of arrow but less dependencies is always better! Big todos are that python-to-arrow-without-pyarrow thing we discussed on the arrow listserv, and upgrading from 0.14.1 to 0.15.0

**Comment by wesm** [comment]
@timkpaine the C API that @pitrou is developing can probably help with this. It would be good to try to validate that for your use case

**Comment by pitrou** [comment]
You can read the Arrow C data interface spec (in draft) here:
https://github.com/apache/arrow/blob/4a0689d0f7837cdae7257dc8d28c41b79d332452/docs/source/format/CProtocol.rst
(from PR https://github.com/apache/arrow/pull/5442)

File: cmake/arrow/CMakeLists.txt (hunk 1 above)
Comment (timkpaine) [reply cmake/arrow/CMakeLists.txt]: fixed

File: cpp/perspective/CMakeLists.txt (hunk 2 above)
Comment (timkpaine) [reply cpp/perspective/CMakeLists.txt]: leaving for now

File: cpp/perspective/CMakeLists.txt (hunk 4 above)
Comment (timkpaine) [reply cpp/perspective/CMakeLists.txt]: fixed

File: cpp/perspective/CMakeLists.txt (hunk 6 above)
Comment (timkpaine) [reply cpp/perspective/CMakeLists.txt]: resolved

**Comment by wesm** [comment]
boost::filesystem is no longer required for a core build (with some components that you don't need disabled). https://github.com/apache/arrow/pull/5611 eliminates Boost altogether from the core build

**Comment by texodus** [comment]
@wesm @pitrou Thanks for the tips - sounds like the new Arrow version has some awesome new features!  Minimizing the dependencies will simplify our WASM build a lot, I'll be looking at upgrading next.

**Pull request merged**